common4j/src/main/com/microsoft/identity/common/java/interfaces/ICommonComponents.java

```diff
@@ -41,4 +42,44 @@
      * Get an encryption manager for storage layer.
      */
     IKeyAccessor getStorageEncryptionManager(@Nullable final ITelemetryCallback telemetryCallback);
+
+    /**
+     * Retrieve a name-value store with a given identifier.
+     *
+     * @param storeName The name of a new KeyValue store.
+     * @param clazz     The class of values in the name value store.
+     * @return a INameValueStorage instance based around data stored with the same storeName.
+     */
+    <T> INameValueStorage<T> getNameValueStore(String storeName, Class<T> clazz);
+
+    /**
+     * Retrieve a name-value store with a given identifier.
+     *
+     * @param storeName The name of a new KeyValue store. May not be null.
+     * @param helper    The key manager for the encryption.  May be null.
+     * @param clazz     The class of values in the name value store. May not be null.
+     * @return a INameValueStorage instance based around data stored with the same storeName.
+     */
+    <T> INameValueStorage<T> getEncryptedNameValueStore(String storeName, IKeyAccessor helper, Class<T> clazz);
+
+    /**
+     * Get a generic encrypted ISharedPreferencesFileManager with a given identifier.
+     *
+     * @param storeName The name of a new KeyValue store. May not be null.
+     * @param helper    The key manager for the encryption.  May not be null.
+     */
+    ISharedPreferencesFileManager getEncryptedFileStore(String storeName, IKeyAccessor helper);
+
+    /**
+     * Get a generic encrypted ISharedPreferencesFileManager with a given identifier.
+     *
+     * @param storeName The name of a new KeyValue store. May not be null.
+     */
+    ISharedPreferencesFileManager getFileStore(String storeName);
+
```

Comment:
It seems the getFileStore methods and getNameValueStore methods kinda serve the same purpose? Also of now Shared Preferences is just Android, so I guess the intention is to rename `ISharedPreferencesFileManager` to something more generic?

Reply:
I also think that instead of having these generic `getNameValueStore` methods on `ICommonComponents`, it might better to have dedicated methods like:

```java
INameValueStorage<String> getTokenCacheStorage(String storeName);

INameValueStorage<String> getDiscoveryFlagStorage(String storeName);

INameValueStorage<String> getAccountManagerLikeStorage(String storeName);

INameValueStorage<String> getTelemetryStorage(String storeName);
```

I feel like this allows us to be able to more easily return different impls of INameValueStorage based on different needs...just going off of storename and class might not be sufficient.

Reply:
I think the Type of the value is the most important thing to consider.  Generally, we should probably be using types richer than String. ;-)

Reply:
I'm not trying to be flip here.  In that case, I'd expect to see something more like:

```
    INameValueStorage<TokenCacheEntry> getPrimaryTokenCache();

    INameValueStorage<TokenCacheEntry> getFociTokenCache();

    INameValueStorage<TokenCacheEntry> getOtherFictionalTokenCache();
```

This swallows the entire idea of these file names.  This would potentially be calling something else that names the resource appropriately and chooses an implementation, but the caller shouldn't care about knowing this magic static name.

Reply:
yes, I think the idea is to get dedicated methods for each of these....in terms of `String` vs `TokenCacheEntry`..I don't know how the `TokenCacheEntry` would internally become a String because our token cache is currently String...but if we can have a way to do it then great

Reply:
This is actually pretty inconvenient in terms of the implementations of INameValueStorage - it doesn't marry well with ISharedPrefrencesFileManager's capabilities in terms of filtering based on key before decryption.  I'm going to take a new run at the problem and see if a different strategy works better.

Reply:
I've renamed this class.